File: sections/35_math_mixed_contest/problems/a_blocked_path/solution.cpp
#include <bits/stdc++.h>
using namespace std;
const long long MOD = 1000000007LL;

long long mod_pow(long long a, long long b) {
    long long r = 1;
    while (b) {
        if (b & 1LL) r = r * a % MOD;
        a = a * a % MOD;
        b >>= 1LL;
    }
    return r;
}

int main() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);
    int h, w, r, c;
    if (!(cin >> h >> w >> r >> c)) return 0;
    int limit = h + w;
    vector<long long> fact(limit + 1, 1), invfact(limit + 1, 1);
    for (int i = 1; i <= limit; i++) fact[i] = fact[i - 1] * i % MOD;
    invfact[limit] = mod_pow(fact[limit], MOD - 2);
    for (int i = limit; i >= 1; i--) invfact[i - 1] = invfact[i] * i % MOD;
    auto C = [&](int n, int k) -> long long {
        if (k < 0 || k > n) return 0;
        return fact[n] * invfact[k] % MOD * invfact[n - k] % MOD;
    };
    long long total = C(h + w - 2, h - 1);
    long long through = C(r + c - 2, r - 1) * C(h - r + w - c, h - r) % MOD;
    cout << (total - through + MOD) % MOD << '\n';
    return 0;
}

File: sections/35_math_mixed_contest/problems/a_blocked_path/solution.py
import sys


MOD = 1_000_000_007


def prepare(n: int) -> tuple[list[int], list[int]]:
    fact = [1] * (n + 1)
    for i in range(1, n + 1):
        fact[i] = fact[i - 1] * i % MOD
    invfact = [1] * (n + 1)
    invfact[n] = pow(fact[n], MOD - 2, MOD)
    for i in range(n, 0, -1):
        invfact[i - 1] = invfact[i] * i % MOD
    return fact, invfact


def comb(n: int, k: int, fact: list[int], invfact: list[int]) -> int:
    if k < 0 or k > n:
        return 0
    return fact[n] * invfact[k] % MOD * invfact[n - k] % MOD


def main() -> None:
    data = list(map(int, sys.stdin.buffer.read().split()))
    if not data:
        return
    h, w, r, c = data
    fact, invfact = prepare(h + w)
    total = comb(h + w - 2, h - 1, fact, invfact)
    through = comb(r + c - 2, r - 1, fact, invfact) * comb(h - r + w - c, h - r, fact, invfact) % MOD
    print((total - through) % MOD)


if __name__ == "__main__":
    main()
